media-source-respec.html

```diff
@@ -786,9 +804,16 @@ <h4><dfn>Attaching to a media element</dfn></h4>
               such as would occur if attempting to use  a [=MediaSource object URL=] from a
               {{DedicatedWorkerGlobalScope}} {{MediaSource}}
               </dt>
-            <dd>Run the <a def-id="media-data-cannot-be-fetched"></a> steps of the <a def-id="resource-fetch-algorithm"></a>'s <a def-id="media-data-processing-steps-list"></a>.</dd>
-            <dt>If {{MediaSource/readyState}} is NOT set to {{ReadyState/""closed""}} 
-            <dd>Run the <a def-id="media-data-cannot-be-fetched"></a> steps of the <a def-id="resource-fetch-algorithm"></a>'s <a def-id="media-data-processing-steps-list"></a>.</dd>
+            <dd>Run the <a def-id="media-data-cannot-be-fetched"></a> steps of the <a
+                def-id="resource-fetch-algorithm"></a>'s <a def-id="media-data-processing-steps-list"></a>.</dd>
+            <dt>If the media provider object is a {{MediaSourceHandle}} or a {{MediaSource}} that is already being used
+              beyond this point as the media provider object in a different {{HTMLMediaElement}}'s
+              <a def-id="resource-fetch-algorithm"></a></dt>
+            <dd>Run the <a def-id="media-data-cannot-be-fetched"></a> steps of the <a
+                def-id="resource-fetch-algorithm"></a>'s <a def-id="media-data-processing-steps-list"></a>.</dd>
+            <dt>If {{MediaSource/readyState}} is NOT set to {{ReadyState/""closed""}}</dt>
+            <dd>Run the <a def-id="media-data-cannot-be-fetched"></a> steps of the <a
+                def-id="resource-fetch-algorithm"></a>'s <a def-id="media-data-processing-steps-list"></a>.</dd>
```

Comment:
Thank you for your analysis and presentation of options. While I think each of those options are viable, I like options 1 and 2 best. They seem to be the simplest in concept and in implementation complexity. For option 1, the resulting MediaError.message could more readily indicate the reason for the error in that case. What are the timing considerations you mentioned might apply to option 2?

Note that initial prototype implementation in Chrome might just forgo some of the complexity involved in allowed sequential re-use of the same MediaSourceHandle (and add that later). The app can always create one or more extra MediaSource instances in the worker and get those handle(s) over to the main thread for readiness to use later as a workaround in the interim.

Reply:
Consider the following two commands running under the same event (i.e. without a stable state between them):
```
video.srcObject = handle;
port.postMessage(null, [handle]);
```

With option 4, the `postMessage()` would fail and the video would keep playing.
With option 3, [[Detached]] is not set to true until the next stable state, and so the `postMessage()` would succeed and the video would error.

Option 2 may be spec'd by marking the handle as in use either in the `srcObject` setter for similar behavior to option 4 or in the "setup worker attachment communication" steps for similar behavior to option 3.

I haven't looked into the when exactly the handle would become unused if unset from `srcObject`, so there may be similar timing issues there for.
```
video.srcObject = null; // previous value was handle
port.postMessage(null, [handle]);
```

Perhaps option 1 is immune to these subtleties.

Reply:
@ https://github.com/w3c/media-source/pull/306#discussion_r846897695 : 

> I haven't looked into the when exactly the handle would become unused if unset from `srcObject`, so there may be similar timing issues there for. [...]

Examining the current html5 spec, upon setting srcObject, synchronous execution of the [media element load algorithm](https://html.spec.whatwg.org/multipage/media.html#media-element-load-algorithm) is done, which has a step that synchronously detaches `MediaSource` if one is the current `assigned media provider object`. So it seems that an implementation could readily know on the main/window thread whether or not a particular `MediaSourceHandle` object is currently assigned to any `HTMLMediaElement` as `srcObject`.

Reply:
Ah, yes.  Thanks for identifying that.
So, if choosing to block transfer while in use (i.e. not option 1), then for some consistency with main-thread MediaSource detach, I guess a `postMessage()` should succeed immediately after a handle is unset through `srcObject`.

Reply:
@karlt, yes, I think I would prefer that.
I discussed this during today's Media Working Group Call, and also have shared this thread on the [media WG mailing list](https://lists.w3.org/Archives/Public/public-media-wg/2022May/0003.html) to get wider input.
Previously, my preference was option 2 (throw exception if currently in use as `srcObject` value), however I have just now observed an issue with that option (perhaps this is what you were trying to point out, too):

```
video.srcObject = handle;
port.postMessage(null, [handle]);
```

While the handle is immediately set to be the media element's `assigned media provider object`, the synchronous portion of this setter (which runs the load algorithm, which then starts the resource selection algorithm) begins the actual attachment in an asynchronous set of steps once the next stable state has been reached. Therefore, the postMessage transfer mechanism, invoked synchronously before that stable state has been reached, does not know that `handle` is in a state preventing transfer. So neither option 3 or 4 can enable option 2 to prevent transfer before that stable state has been reached.

Does this mean option 2 is infeasible? Maybe, unless we update the srcObject setter spec for when a `MediaSourceHandle` is set as the assigned media provider object in the synchronous section of the steps following that assignment (for example, in the load algorithm). If we did that, then we could reliably know when the handle is assigned (or not) at any attempt to transfer it. (Note, unsetting the handle as assigned media provider object is already synchronously running MSE detachment as I noted earlier today.)

Overall, my preference is to lean towards giving application synchronous awareness of failure to transfer due to handle being in-use as srcObject already. I think this would help web authors avoid having to debug additional asynchronous error transitions that are already quite overloaded with various kinds of HTMLMediaElement and MSE errors.

Reply:
I agree that, if choosing to block transfer while in use, then marking the handle as in use synchronously from the `srcObject` setter would produce the behavior with probably the least surprise.

> @ [#306 (comment)](https://github.com/w3c/media-source/pull/306#discussion_r846897695) :
> Examining the current html5 spec, upon setting srcObject, synchronous execution of the [media element load algorithm](https://html.spec.whatwg.org/multipage/media.html#media-element-load-algorithm) is done, which has a step that synchronously detaches `MediaSource` if one is the current `assigned media provider object`.

FWIW I was assuming this detach was for the old/previous `assigned media provider object`.  Perhaps that was intended, but that's not what the spec currently says.  The `srcObject` setter ["must set the element's `assigned media provider object` to the new value, and then invoke the element's media element load algorithm."](https://html.spec.whatwg.org/multipage/media.html#dom-media-srcobject), so the media element load algorithm detaches the new `MediaSource` object.

Still a synchronous release of the handle from in use when cleared from `srcObject` would be the least surprise behavior here.